eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala

```diff
@@ -0,0 +1,75 @@
+package fr.acinq.eclair.io
+
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.scalacompat.{ByteVector32, Satoshi}
+import fr.acinq.eclair.channel.{ChannelConfig, LocalParams, SupportedChannelType}
+import fr.acinq.eclair.io.Peer.{ConnectedData, OutgoingMessage, SpawnChannelNonInitiator}
+import fr.acinq.eclair.wire.protocol.{Error, OpenChannel, OpenDualFundedChannel}
+import fr.acinq.eclair.{AcceptOpenChannel, InterceptOpenChannelPlugin, InterceptOpenChannelReceived, InterceptOpenChannelResponse, RejectOpenChannel}
+
+import scala.concurrent.duration.FiniteDuration
+
+// Child actor of a Peer that handles the interactions with an InterceptOpenChannelPlugin:
+//  - sends OpenChannelReceived to the plugin
+//  - waits for its response and translates it to the Peer data format (either SpawnChannelNonInitiator or an outgoing Error)
+//  - reject the channel open after a timeout (configurable in eclair.conf)
+//
+//  Note: we don't fully trust plugins to be correctly implemented, and we need to respond to our peer even if the plugin fails to tell us what to do
+
+object OpenChannelInterceptor {
+  def apply(replyTo: ActorRef[Any], plugin: InterceptOpenChannelPlugin, timeout: FiniteDuration, connectedData: ConnectedData, temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount: Satoshi, open: Either[OpenChannel, OpenDualFundedChannel], channelType: SupportedChannelType): Behavior[Command] = {
+    Behaviors.setup { context =>
+      Behaviors.withTimers { timers =>
+        timers.startSingleTimer(PluginTimeout, timeout)
+        new OpenChannelInterceptor(replyTo, plugin, connectedData, temporaryChannelId, localParams, fundingAmount, open, channelType, context).start()
+      }
+    }
+  }
+
+  // @formatter:off
+  sealed trait Command
+  case object PluginTimeout extends Command
+  case class WrappedOpenChannelResponse(pluginResponse: InterceptOpenChannelResponse) extends Command
+  // @formatter:on
+}
+
+private class OpenChannelInterceptor(replyTo: ActorRef[Any], plugin: InterceptOpenChannelPlugin, connectedData: ConnectedData, temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount: Satoshi, open: Either[OpenChannel, OpenDualFundedChannel], channelType: SupportedChannelType, context: ActorContext[OpenChannelInterceptor.Command]) {
+
+  import OpenChannelInterceptor._
+
+  private def start(): Behavior[Command] = {
+    val pluginResponseAdapter = context.messageAdapter[InterceptOpenChannelResponse](WrappedOpenChannelResponse)
+    plugin.getOpenChannelInterceptor ! InterceptOpenChannelReceived(pluginResponseAdapter, open, temporaryChannelId, localParams, fundingAmount)
+
+    Behaviors.receiveMessage {
+      case PluginTimeout =>
+        context.log.error(s"plugin ${plugin.name} timed out while intercepting open channel")
+        replyTo ! OutgoingMessage(Error(temporaryChannelId, "plugin timeout"), connectedData.peerConnection)
```

Comment:
There is a subtlety in the way you use the peer connection: the `Peer` will drop the message if the current connection doesn't match the one from this actor. This is actually ok, because if we disconnected before sending `accept_channel`, the channel opening flow is cancelled according to the lightning spec, so we shouldn't send those messages on the new connection. This is worth a comment!

Reply:
Should we also include `peerConnection` in the `SpawnChannelNonInitiator` message and update `Peer` to ignore it if `peerConnection` has changed?  

Otherwise I think we could be in a situation where the remote peer disconnects before we send `accept_channel` and according to the spec the remote peer must send a new `ChannelOpen` which we could accidentally reply to with an `accept_channel` from their previous `ChannelOpen` from before they disconnected.

Reply:
Updated comment in [3ee062b](https://github.com/ACINQ/eclair/pull/2552/commits/3ee062ba7cd78aa5e550e7917dd3709ea50f51a3). But may need to also update  how `SpawnChannelNonInitiator` is handled by `Peer` to live up to the comment.